File: Union Find/0582.Kill Process/Solution.java
class Solution {
    int[] p;
    int kill;

    public List<Integer> killProcess(List<Integer> pid, List<Integer> ppid, int kill) {
        int n = pid.size();
        this.kill = kill;
        p = new int[50010];
        for (int i = 0; i < p.length; i++) {
            p[i] = i;
        }
        for (int i = 0; i < n; i++) {
            union(pid.get(i), ppid.get(i));
        }
        List<Integer> ans = new ArrayList<>();
        for (int id : pid) {
            if (find(id) == kill) {
                ans.add(id);
            }
        }
        return ans;
    }

    private int find(int x) {
        if (x == kill) {
            return x;
        }
        if (p[x] != x) {
            p[x] = find(p[x]);
        }
        return p[x];
    }

    private void union(int i1, int i2) {
        int r1 = find(i1), r2 = find(i2);
        if (r1 != r2) {
            p[r1] = r2;
        }
    }
}

File: Union Find/0582.Kill Process/Solution.py
class Solution:
    def killProcess(self, pid: List[int], ppid: List[int], kill: int) -> List[int]:
        p = [i for i in range(50010)]

        def find(x):
            if x == kill:
                return x
            if p[x] != x:
                p[x] = find(p[x])
            return p[x]

        def union(i1, i2):
            r1, r2 = find(i1), find(i2)
            if r1 != r2:
                p[r1] = r2

        n = len(pid)
        for i in range(n):
            union(pid[i], ppid[i])
        ans = []
        for v in pid:
            if find(v) == kill:
                ans.append(v)
        return ans